src/tracing.h

```diff
@@ -15,16 +15,31 @@ extern struct raft_tracer NoopTracer;
 extern struct raft_tracer StderrTracer;
 
 /* Emit a debug message with the given tracer. */
-#define Tracef(TRACER, ...)                                    \
-    do {                                                       \
-        if (UNLIKELY(TRACER != NULL && TRACER->emit != NULL && \
-                     TRACER->enabled)) {                       \
-            static char _msg[1024];                            \
-            snprintf(_msg, sizeof _msg, __VA_ARGS__);          \
-            TRACER->emit(TRACER, __FILE__, __LINE__, _msg);    \
-        }                                                      \
+#define Tracef0(TRACER, LEVEL, ...)                                            \
+    do {                                                                       \
+        if (UNLIKELY(TRACER != NULL && TRACER->emit != NULL &&                 \
+                     TRACER->enabled)) {                                       \
+            char _msg[1024];                                                   \
+            snprintf(_msg, sizeof _msg, __VA_ARGS__);                          \
+            TRACER->emit(TRACER, __FILE__, __LINE__, __func__, (LEVEL), _msg); \
+        }                                                                      \
     } while (0)
 
+enum raft_trace_level {
+    TRACE_NONE,
+    TRACE_ALWAYS,
+    TRACE_FATAL,
+    TRACE_ERROR,
+    TRACE_WARN,
+    TRACE_NOTICE,
+    TRACE_INFO,
+    TRACE_DEBUG,
+    TRACE_CALL,
+    TRACE_NR
+};
```

Comment:
Do we need this many levels? I feel like I'm not going to know which one to choose when I'm writing code without some sense of what the distinctions are.

Reply:
```
	/**
	 * represents an invalid trace level
	 */
	M0_NONE   = 0,

	/**
	 * special level, intended for
	 * printf-style debugging, should be removed from code before committing
	 */
	M0_ALWAYS = 1 << 0,

	/** system is not able to continue performing it's basic function */
	M0_FATAL  = 1 << 1,
	/**
	 * resource unavailable, no connectivity,
	 * incorrect value, etc.
	 */
	M0_ERROR  = 1 << 2,
	/**
	 * condition which requires some special
	 * treatment, something which not happens normally.
	 */
	M0_WARN   = 1 << 3,
	/** normal but significant code path */
	M0_NOTICE = 1 << 4,
	/** information about current system's state */
	M0_INFO   = 1 << 5,
	/**
	 * lower-level information to debug and analyse
	 * incorrect behavior
	 */
	M0_DEBUG  = 1 << 6,

	/**
	 * special level, used only to trace function calls
	 */
	M0_CALL   = 1 << 7,
```
We can try to narrow down the list.

Reply:
let's remove `ALWAYS`. It's too early for this class of logs.

Reply:
`-NOTICE` and `-CALL`. In total:

```
	/**
	 * represents an invalid trace level
	 */
	M0_NONE  ,

	/** system is not able to continue performing it's basic function */
	M0_FATAL ,
	/**
	 * resource unavailable, no connectivity,
	 * incorrect value, etc.
	 */
	M0_ERROR  ,
	/**
	 * condition which requires some special
	 * treatment, something which not happens normally.
	 */
	M0_WARN  ,
	/** information about current system's state */
	M0_INFO  ,
	/**
	 * lower-level information to debug and analyse
	 * incorrect behavior
	 */
	M0_DEBUG  ,
```

Reply:
Thanks, this list is much easier for me to understand. I like that there's a distinct level for transient debugging stuff! Could you add those comments to your PR for posterity?

Reply:
done